internal/cli/cmd/auth/authorize/cached_provider.go

```diff
@@ -0,0 +1,324 @@
+
+This program is distributed in the hope that it will be useful
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package authorize
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/99designs/keyring"
+	"github.com/mitchellh/go-homedir"
+	"github.com/pkg/errors"
+
+	"github.com/apecloud/kubeblocks/internal/cli/cmd/auth/authorize/authenticator"
+)
+
+const (
+	defaultConfigPath = "~/.config/kueblocks"
+	userInfoFile      = "user_info.json"
+	tokenFile         = "token.json"
+
+	keyringKey     = "token"
+	keyringService = "kueblocks"
+	keyringLabel   = "KUBEBLOCKS CLI"
+
+	fileMode = 0o600
+)
+
+type KeyringCached struct {
+	key     string
+	valid   bool
+	keyring keyring.Keyring
+}
+
+func (k *KeyringCached) isValid() bool {
+	return k.valid
+}
+
+func (k *KeyringCached) get() ([]byte, error) {
+	item, err := k.keyring.Get(k.key)
+	if err != nil {
+		return nil, err
+	}
+	return item.Data, nil
+}
+
+func (k *KeyringCached) set(data []byte) error {
+	return k.keyring.Set(keyring.Item{
+		Key:   k.key,
+		Data:  data,
+		Label: keyringLabel,
+	})
+}
+
+func (k *KeyringCached) remove() error {
+	return k.keyring.Remove(k.key)
+}
+
+type FileCached struct {
+	tokenFilename     string
+	userInfoFilename  string
+	defaultConfigPath string
+}
+
+type KeyringCachedTokenProvider struct {
+	keyringCached KeyringProvider
+	fileCached    FileCached
+}
+
+func NewKeyringCachedTokenProvider(keyringCached *KeyringProvider) *KeyringCachedTokenProvider {
+	fileCached := FileCached{
+		tokenFilename:     tokenFile,
+		userInfoFilename:  userInfoFile,
+		defaultConfigPath: defaultConfigPath,
+	}
+
+	if keyringCached == nil {
+		defaultKeyring, isValid := defaultKeyring()
+		return &KeyringCachedTokenProvider{
+			keyringCached: &KeyringCached{
+				key:     keyringKey,
+				valid:   isValid,
+				keyring: defaultKeyring,
+			},
+			fileCached: fileCached,
+		}
+	}
+
+	return &KeyringCachedTokenProvider{
+		keyringCached: *keyringCached,
+		fileCached:    fileCached,
+	}
+}
+
+func defaultKeyring() (keyring.Keyring, bool) {
+	k, err := keyring.Open(keyring.Config{
+		AllowedBackends: []keyring.BackendType{
+			keyring.SecretServiceBackend,
+			keyring.KWalletBackend,
+			keyring.KeychainBackend,
+			keyring.WinCredBackend,
+		},
+		ServiceName:              keyringService,
+		KeychainTrustApplication: true,
+		KeychainSynchronizable:   true,
+	})
+
+	if err != nil {
+		return nil, false
+	}
+	return k, true
+}
+
+func (k *KeyringCachedTokenProvider) GetTokens() (*authenticator.TokenResponse, error) {
+	if !k.keyringCached.isValid() {
+		token, tokenErr := k.fileCached.readToken()
+		if os.IsNotExist(tokenErr) {
+			return nil, nil
```

Comment:
return error should be set

Reply:
The reason why I did not return an error here is: I distinguished between reading token errors and token non-existence. When reading token errors, an error is reported, and when tokens do not exist, log in again.